src/serac/physics/materials/thermal_functional_material.hpp

```diff
@@ -0,0 +1,239 @@
+// Copyright (c) 2019-2021, Lawrence Livermore National Security, LLC and
+// other Serac Project Developers. See the top-level LICENSE file for
+// details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+/**
+ * @file thermal_functional_material.hpp
+ *
+ * @brief The material and load types for the thermal functional physics module
+ */
+
+#pragma once
+
+#include "serac/numerics/functional/functional.hpp"
+
+/// ThermalConductionFunctional helper structs
+namespace serac::Thermal {
+
+/// Linear isotropic thermal conduction material model
+struct LinearIsotropicConductor {
+  /// Density
+  double density_ = -1.0;
+
+  /// Specific heat capacity
+  double specific_heat_capacity_ = -1.0;
+
+  /// Constant isotropic thermal conductivity
+  double conductivity_ = -1.0;
+
+  /**
+   * @brief Function defining the thermal flux (constitutive response)
+   *
+   * @tparam T1 type of the temperature (e.g. tensor or dual type)
+   * @tparam T2 type of the temperature gradient (e.g. tensor or dual type)
+   * @param temperature_gradient Gradient of the temperature (du_dx)
+   * @return The thermal flux of the material model
+   */
+  template <typename T1, typename T2>
+  SERAC_HOST_DEVICE T2 operator()(T1& /* temperature */, T2& temperature_gradient) const
+  {
+    SLIC_ASSERT_MSG(conductivity_ > 0.0,
+                    "Conductivity must be positive in the linear isotropic conductor material model.");
+
+    return -1.0 * conductivity_ * temperature_gradient;
+  }
+
+  /**
+   * @brief The density (mass per volume) of the material model
+   *
+   * @tparam T1 type of the position variable
+   * @return The density
+   */
+  template <typename T1>
+  SERAC_HOST_DEVICE double density(T1& /* x */) const
+  {
+    SLIC_ASSERT_MSG(density_ > 0.0, "Density must be positive in the linear isotropic conductor material model.");
+
+    return density_;
+  }
+
+  /**
+   * @brief The specific heat capacity (heat capacity per unit mass) of the material model
+   *
+   * @tparam T1 Type of the position variable
+   * @tparam T2 Type of the temperature variable
+   */
+  template <typename T1, typename T2>
+  SERAC_HOST_DEVICE double specificHeatCapacity(T1& /* x */, T2& /* temperature */) const
+  {
+    SLIC_ASSERT_MSG(specific_heat_capacity_ > 0.0,
+                    "Specific heat capacity must be positive in the linear isotropic conductor material model.");
+
+    return specific_heat_capacity_;
+  }
+};
+
+/**
+ * @brief Linear anisotropic thermal material model
+ *
+ * @tparam dim Spatial dimension
+ */
+template <int dim>
+struct LinearConductor {
+  /// Density
+  double density_ = -1.0;
+
+  /// Specific heat capacity
+  double specific_heat_capacity_ = -1.0;
+
+  /// Constant isotropic thermal conductivity
+  tensor<double, dim, dim> conductivity_ = {-1.0};
```

Comment:
these seem like unusual choices for default values

I'm also not sure what `tensor<double, dim, dim> conductivity_ = {-1.0};` actually does

Reply:
I wanted a default value that would fail if uninitialized. I assume the `= double` would just fill all of the entries with the single value, but I admit I have not checked.

Reply:
the tensor class doesn't define any constructors, so it initializes kind of like a multidimensional array in C++. That means the following code 
```cpp
std::cout << serac::tensor< double, 3, 3>{-1.0};
```
will print `{{-1, 0, 0}, {0, 0, 0}, {0, 0, 0}}`, in much the same way as https://godbolt.org/z/5hzfzfvb3

Reply:
I put a reasonable default of `1.0` in for now.